gcloud-java-core/src/main/java/com/google/cloud/BatchResult.java

```diff
@@ -51,27 +56,39 @@ public T get() throws E {
   }
 
   /**
-   * Registers a callback for the batch operation.
+   * Adds a callback for the batch operation. If the batch has been completed already, the callback
+   * will be invoked immediately.
    */
   public void notify(Callback<T, E> callback) {
-    // todo(mderka) implement
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (!submitted) {
+      toBeNotified.add(callback);
+    } else if (error != null) {
+      callback.error(error);
+    } else {
+      callback.success(result);
+    }
   }
 
   /**
-   * Sets an error and makes this submitted.
+   * Sets an error and makes this submitted. Notifies all callbacks.
    */
   protected void error(E error) {
-    this.error = error;
+    this.error = checkNotNull(error);
     this.submitted = true;
+    for (Callback<T, E> callback : toBeNotified) {
+      callback.error(error);
+    }
   }
 
   /**
-   * Sets a result and makes this submitted.
+   * Sets a result and makes this submitted. Notifies all callbacks.
    */
   protected void success(T result) {
-    this.result = result;
+    this.result = checkNotNull(result);
```

Comment:
We should probably add an associated test that would have failed as the code is written here.